### ISSUE TITLE
Queries wait a fixed 500ms when mongodb stopped. Only on second and subsequent queries.

### ISSUE THREAD

**Issue by cb-msales** [mongodb/mongo-php-driver#592]
### Description

I experience a 500ms wait time in case mongodb server is not running / stopped.
But weird thing is that first request fails in < 1 ms but subsequent requests all take at least 500ms.
It doesn't matter if the php script waits or the request just takes that time. 500ms seems to be some fixed value here.

Please check my PHP script below and try by yourself. No MongoDB server required.
Feel free to add timeout parameters but I didn't had any luck with them.

### Environment

/usr/local/etc/php/7.1/conf.d/ext-mongodb.ini,
mongodb
mongodb support => enabled
mongodb version => 1.2.2
mongodb stability => stable
libmongoc version => 1.5.0
libbson version => 1.5.0
mongodb.debug => no value => no value

### Test Script

```
<?php

function fireAndMeasureTime(\MongoDB\Driver\Manager $manager, $number) {
    $startMicro = microtime(true);
    try {
        $result = $manager->executeCommand('local', new \MongoDB\Driver\Command(['ping' => 1]));
    } catch(\Throwable $exception) {
        echo $number.". query failed with " . $exception->getMessage() . "\n";
    }
    echo $number.". query took " . (microtime(true) - $startMicro)*1000 . "ms\n";
}

// Assuming you have no local mongodb instance running there.
$manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:3000');

fireAndMeasureTime($manager, 1); // fails very fast:  < 1ms

// Try to connect to a port running nothing.
$manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:12345');

fireAndMeasureTime($manager, 2); // fails very fast too: < 1ms

fireAndMeasureTime($manager, 3); // Second request on same port failes after 500ms.

usleep(200*1000); // it doesn't matter where the 500ms above are spent, so spend some in php to show effect.

fireAndMeasureTime($manager, 4); // ~300ms only because we waited 200 ms before

// switch port back and forth
$manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:3000');
$manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:12345');

fireAndMeasureTime($manager, 5); // still slow, as if the 500ms wait time is required.

// switch back to first connection (has been used more than 500ms ago)
$manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:3000');

fireAndMeasureTime($manager, 6); // fast again, 500ms wait window is over on this connection.

```

Outpout:

```
1. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:3000']
1. query took 0.65803527832031ms
2. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:12345']
2. query took 0.1680850982666ms
3. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:12345']
3. query took 505.54895401001ms
4. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:12345']
4. query took 297.63507843018ms
5. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:12345']
5. query took 500.42200088501ms
6. query failed with No suitable servers found (`serverSelectionTryOnce` set): [connection refused calling ismaster on '127.0.0.1:3000']
6. query took 0.04887580871582ms
```

### Expected and Actual Behavior

Actual: See output above, long wait time before errors happen.

Expected: Always fail immediately or fail after some configurable timeout < 500ms

**Comment by jmikola** [comment]
The 500ms delay corresponds to [`minHeartbeatFrequencyMS`](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#minheartbeatfrequencyms), which is the minimum interval that the driver will recheck a server according to the cross-driver [Server Discovery and Monitoring specification]((https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst).

Based on the driver's [connection persistence](http://php.net/manual/en/mongodb.connection-handling.php) and Manager constructor arguments in your script, two persistent clients are being created and re-used throughout the script. This would be evident if you examined the [`mongodb.debug`](http://php.net/manual/en/mongodb.configuration.php#ini.mongodb.debug) logs for the script as well.

The first time you execute `fireAndMeasureTime()` for each persisted client, the driver starts from a blank slate and attempts to establish new connections to each host in the URI to discovery the topology. This results in immediate "connection refused" errors.

The later Manager instances re-use persisted clients, which remember when they last attempted to communicate with each host in the URI. When `executeCommand()` invokes server selection to find a primary server, that selection blocks until the driver can re-examine the topology.

The internal logic for this behavior is in [`mongoc_topology_select_server_id()`](https://github.com/mongodb/mongo-c-driver/blob/1.5.0/src/mongoc/mongoc-topology.c#L507). When the [first selection fails to find a server](https://github.com/mongodb/mongo-c-driver/blob/1.5.0/src/mongoc/mongoc-topology.c#L584), the topology is marked as "stale" to ensure that subsequent selection will [do a scan before attempting selection again](https://github.com/mongodb/mongo-c-driver/blob/1.5.0/src/mongoc/mongoc-topology.c#L552). That second scan is where the `minHeartbeatFrequencyMS` delay comes in, as the SDAM specification avoids scanning the same host within a lesser time frame.

> ```
> // switch port back and forth
> $manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:3000');
> $manager = new \MongoDB\Driver\Manager('mongodb://127.0.0.1:12345');
> ```

For the record, the first assignment in this snippet has no effect.

As a side note, your report mentions that you're using 1.2.2. In the future, please attempt to reproduce issues using the latest stable release ([1.2.9](https://pecl.php.net/package/mongodb/1.2.9) at present). It wouldn't have made a difference in this case, but it can avoid back and forth had this been an actual bug that was fixed since.

**Comment by cb-msales** [comment]
Thanks you for the good explanation. I understand why there is 500ms time frame before the next server selection will start.

What I'm not clear about yet is why the server selection would block the command instead of just making it fail quickly, as at the time the command is issue, there is clearly no usable server available.

Some kind of retry mechanism could be implemented in application code, after all application logic only knows if it is better to wait and retry or ignore and move on.

Also from a developer point of view there are timeouts to be configured and in case all timeouts are set to let's say 50ms it is unexpected to have any command blocking for 500ms.

Is there a way to work around this wait time, e.g. adding a option to query like "skipServerDiscoveryWait"?

**Comment by cb-msales** [comment]
Is [waitQueueTimeoutMS](https://docs.mongodb.com/manual/reference/connection-string/) what I'm looking for?

According to c driver it is [not implemented](https://github.com/mongodb/mongo-c-driver/blob/1.5.0/doc/mongoc_uri_t.page#L96).

**Comment by jmikola** [comment]
None of the [connection pool URI options](https://docs.mongodb.com/manual/reference/connection-string/#connection-pool-options) apply to the PHP driver, so we can rule out `waitQueueTimeoutMS`.

Thinking about this some more, I wonder if `serverSelectionTimeoutMS` may be relevant to coerce an earlier failure. Although the default value is 30000, I tested a smaller amount of 100 and found that server selection still blocks for 500 milliseconds.

I'll follow up with the C driver team to ask if this is expected behavior, or if the spec may need to be revised for our single-threaded mode.

**Comment by jmikola** [comment]
After speaking with the C driver team about this, I created [CDRIVER-2160](https://jira.mongodb.org/browse/CDRIVER-2160) to request some changes to the logic for single-threaded topology scanning. Ideally, libmongoc should determine whether any nodes are even eligible to be scanned (taking into account the 5-second `cooldownMS` limit after a network error) before deciding to sleep for `minHeartbeatFrequencyMS`.

> Thinking about this some more, I wonder if `serverSelectionTimeoutMS` may be relevant to coerce an earlier failure. Although the default value is 30000, I tested a smaller amount of 100 and found that server selection still blocks for 500 milliseconds.

I mention this in the aforementioned ticket, but this is due to the default try-once behavior (configurable via the `serverSelectionTryOnce` URI option).

**Comment by jmikola** [comment]
[CDRIVER-2160](https://jira.mongodb.org/browse/CDRIVER-2160) is being addressed in libmongoc 1.10, which will be included in the forthcoming 1.5.0 release of the PHP driver. We don't have a specific release date set for 1.5.0, but please watch the project for updates.

Closing this out as there are no other action items remaining in this ticket.
